optuna/integration/mlflow.py

```diff
@@ -239,14 +247,37 @@ def _set_tags(self, trial: optuna.trial.FrozenTrial, study: optuna.study.Study)
         # This sets the tags for MLflow.
         mlflow.set_tags(tags)
 
-    def _log_metric(self, value: Union[float, None]) -> None:
-        """Log the trial result as metric to MLflow.
+    def _log_metrics(self, values: List[Union[float, None]]) -> None:
+        """Log the trial results as metrics to MLflow.
 
         Args:
-            value: Result of trial.
+            values: Results of a trial.
         """
-        trial_value = value if value is not None else float("nan")
-        mlflow.log_metric(self._metric_name, trial_value)
+
+        if isinstance(self._metric_name, list):
```

Comment:
When I set a tuple of strings, the following error occurred. In my understanding, `self._metric_name` is not necessarily a list. How about loose the type to `typing.Sequence`? See also https://github.com/optuna/optuna/blob/cbae80476c15b6d39e1d8851dc6a501c63c3ca92/optuna/study/study.py#L496 https://github.com/optuna/optuna/blob/cbae80476c15b6d39e1d8851dc6a501c63c3ca92/optuna/study/_optimize.py#L271

```console
mlflow.exceptions.MlflowException: Invalid metric name: '('foo', 'bar')_0'. Names] may only contain alphanumerics, underscores (_), dashes (-), periods (.), spaces ( ), and slashes (/).
```

Reply:
I will relax type of `metric_name` in constructor, but internally we probably should keep checking for instance of `List` and `Tuple`, since checking just for `Sequence` will also pass strings.

Reply:
Note: also TODO in `WeightsAndBiasesCallback`

Reply:
Addressed in b3ee6bb69a30a97cb930462735b014c9f104cff1. At the end, constructor is relaxed to `Sequence`, and internally we are just checking for instance of `str`. This should enable user to ducktype any sequence safely. Also added test case covering error you have reported.

Reply:
> constructor is relaxed to `Sequence`, and internally we are just checking for instance of `str`.

Related to this design, I added a suggestion in https://github.com/optuna/optuna/pull/2863#discussion_r695407397.